### PR TITLE
tools/packaging: Add option to build a QAT kernel

### LINKED ISSUE #2578: tools/packaging: Add option to build QAT kernels
Add an option to allow user build a kernel that supports Intel QAT

### PR DESCRIPTION
**Pull request by devimc**
Add -q option to allow users build a guest kernel that supports
Intel QuickAssist Technology.

fixes #2578

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

**Comment by devimc** [comment]
cc @eadamsintel @GabyCT

**Review by GabyCT — APPROVED** [review]

**Comment by dborquez** [comment]
@devimc thank you, lgtm

**Review by dborquez — APPROVED** [review]

File: tools/packaging/kernel/build-kernel.sh [hunk 1]
```diff
@@ -43,10 +43,12 @@ experimental_kernel="false"
 force_setup_generate_config="false"
 #GPU kernel support
 gpu_vendor=""
-#Confidential guest type 
+#Confidential guest type
```
Comment (fidencio) [inline tools/packaging/kernel/build-kernel.sh:46]: Not exactly related to this PR. :-)

**Review by fidencio — APPROVED** [review]
lgtm, thanks @devimc!

Today is a holiday in Camb^W the US, would you mind holding this one till @eadamsintel has a chance to take a look at it?

**Comment by devimc** [comment]
@fidencio sure, I can wait for @eadamsintel , thanks

**Comment by eadamsintel** [comment]
I would like @mythi to take a look as well.  After the lab move I don't have access to my systems to test this out and another urgent task got accelerated while I was out. This looks like it is building a rootfs as well. Is that correct?  The early comments I have on this is that 1) Eventually we want to migrate to upstream driver instead of OOT driver which should make this process easier once the detailed steps have been figured out 2) There is already a Dockerfile to build the OOT QAT driver and rootfs so we need to decide if it is worth maintaining both 3) The QAT download link will change once a new version of the OOT driver is out.  I have tried to get a permanent link but the nature of downloadmirror.intel.com makes that hard.

**Comment by devimc** [comment]
@eadamsintel 

> I would like @mythi to take a look as well. 

yeah, sounds good

> This looks like it is building a rootfs as well. Is that correct? 

no, this builds the kernel and QAT modules, I will raise a PR later to build a rootfs image with the QAT modules

>  1) Eventually we want to migrate to upstream driver instead of OOT driver which should make this process easier once the detailed steps have been figured out 

great! this can simplify the things a lot, do you know when this will happen?

> 2) There is already a Dockerfile to build the OOT QAT driver and rootfs so we need to decide if it is worth maintaining both

could you point me to that file?

> 3) The QAT download link will change once a new version of the OOT driver is out. I have tried to get a permanent link but the nature of downloadmirror.intel.com makes that hard.

I have added a new section in the version.yaml file for the QAT driver,  scripts will get the QAT version from that file, bump to a new version should be easier.

**Comment by eadamsintel** [comment]
> > 1. Eventually we want to migrate to upstream driver instead of OOT driver which should make this process easier once the detailed steps have been figured out
> 
> great! this can simplify the things a lot, do you know when this will happen?

I would say Q3 or maybe early Q4 the instructions should be figured out. The kernel and rootfs should be much easier to build and use but it might require some tweaks of the QEMU parameters.

> 
> > 1. There is already a Dockerfile to build the OOT QAT driver and rootfs so we need to decide if it is worth maintaining both
> 
> could you point me to that file?

https://github.com/kata-containers/kata-containers/tree/main/tools/osbuilder/dockerfiles/QAT

> 
> I have added a new section in the version.yaml file for the QAT driver, scripts will get the QAT version from that file, bump to a new version should be easier.

Yes and the 1.7 OOT driver has not changed for a few months now so it wouldn't surprise me if this gets updated soon. I am on the subscribe list for it so I can update the QAT Dockerfile version.  I will keep an eye out for those announcements so both locations can be updated.

**Comment by mythi** [comment]
> 2) There is already a Dockerfile to build the OOT QAT driver and rootfs so we need to decide if it is worth maintaining both

This appoach would be my preference when the OOT driver is needed. @devimc #2578 does not specify the use case. Are you just looking to add QAT validation in Kata?

**Comment by devimc** [comment]
@mythi  
> @devimc #2578 does not specify the use case. Are you just looking to add QAT validation in Kata?

yes, that's the goal

**Comment by mythi** [comment]
@devimc we should set up the validation without the OOT drivers and go with the approach @eadamsintel mentioned. In principle that means: enable vIOMMU, bind the QAT VF to vfio-pci inside the guest and run something like `intel/crypto-perf:devel` container with it.

**Comment by devimc** [comment]
@eadamsintel  and I have discussed this offline,  let's wait for the upstream QAT releases and rework this, we'll be using the Dockerfile to build the kernel and rootfs in our CI

thanks @mythi  @eadamsintel

**Pull request closed without merging**